docs/docs/concepts/triggers/KeySpace_Triggers.md

```diff
@@ -0,0 +1,205 @@
+    if (client.call("type", data.key) != "hash") {
+        // key is not a has, do not touch it.
+        return;
+    }
+    // get the current time in ms
+    var curr_time = client.call("time")[0];
+    // set '__last_updated__' with the current time value
+    client.call('hset', data.key, '__last_updated__', curr_time);
+});
+```
+
+Argument Description:
+
+* consumer - the consumer name.
+* prefix - the prefix of keys we want to fire the trigger on
+* callback - the callback to invoke. Following the same rules of [Sync and Async invocation](../Sync_Async.md). The callback will be invoke only on primary shard.
+
+Run example:
+
+```bash
+127.0.0.1:6379> hset h x 1
+(integer) 1
+127.0.0.1:6379> hgetall h
+1) "x"
+2) "1"
+3) "__last_updated__"
+4) "1658390831"
+127.0.0.1:6379> hincrby h x 1
+(integer) 2
+127.0.0.1:6379> hgetall h
+1) "x"
+2) "2"
+3) "__last_updated__"
+4) "1658390910"
+```
+
+The `data` argument which pass to the consumer callback are in the following format:
+
+```json
+{
+    "event": "<the event name that fired the trigger>",
+    "key": "<key name that the event was fired on as String>",
+    "key_raw": "<key name that the event was fired on as ArrayBuffer>"
+}
+```
+
+Notice that `key` field is given only if the key can be decoded as `String`, otherwise the value will be `null`.
+
+We can observe the trigger information using `TFUNCTION LIST` command:
+
+```bash
+127.0.0.1:6379> TFUNCTION list vvv
+1)  1) "engine"
+    2) "js"
+    3) "api_version"
+    4) "1.0"
+    5) "name"
+    6) "foo"
+    7) "pending_jobs"
+    8) (integer) 0
+    9) "user"
+    10) "default"
+    11) "functions"
+   12) (empty array)
+   13) "keyspace_triggers"
+   14) (empty array)
+   15) "stream_triggers"
+   16) 1)  1) "name"
+           2) "consumer"
+           3) "num_triggered"
+           4) (integer) 2
+           5) "num_finished"
+           6) (integer) 2
+           7) "num_success"
+           8) (integer) 1
+           9) "num_failed"
+          10) (integer) 1
+          11) "last_error"
+          12) "TypeError: redis.call is not a function"
+          13) "last_exection_time"
+          14) (integer) 0
+          15) "total_exection_time"
+          16) (integer) 0
+          17) "avg_exection_time"
+          18) "0"
+```
+
+## Triggers Guarantees
+
+If the callback pass to the trigger is a `JS` function (not a Coroutine), it is guarantee that the callback will be invoke atomically along side the operation that cause the trigger, i.e all client will see the data only after the callback has finished. In addition, it is guarantee that the effect of the callback will be replicated to the replication and the AOF in a `multi/exec` block together with the command that fired the trigger.
+
+If the callback is a Coroutine, it will be executed in the background and there is not guarantees on where or if it will be executed. The guarantees are the same as describe on [Sync and Async invocation](../Sync_Async.md).
```

Comment:
```suggestion
If the callback function passed to the trigger is a `JS` function (not a Coroutine), it is guaranteed that the callback will be invoked atomically along side the operation that caused the trigger; meaning all clients will see the data only after the callback has completed. In addition, it is guaranteed that the effect of the callback will be replicated to the replica and the AOF in a `multi/exec` block together with the command that fired the trigger.

If the callback is a Coroutine, it will be executed in the background and there is no guarantee on where or if it will be executed. The guarantees are the same as described on [Sync and Async invocation](../Sync_Async.md).
```